mysql/datadog_checks/mysql/execution_plans.py

```diff
@@ -307,32 +312,106 @@ def _collect_execution_plans(self, db, tags, options, min_collection_interval):
         statsd.gauge("dd.mysql.collect_execution_plans.seen_statement_plan_sigs", len(seen_statement_plan_sigs),
                      tags=instance_tags)
 
-    def _run_explain(self, cursor, statement, schema, instance_tags):
-        # TODO: cleaner query cleaning to strip comments, etc.
-        if statement.strip().split(' ', 1)[0].lower() not in VALID_EXPLAIN_STATEMENTS:
-            return
+    def _attempt_explain(self, cursor, statement, schema):
+        """
+        Tries the available methods used to explain a statement for the given schema. If a non-retryable
+        error occurs (such as a permissions error), then statements executed under the schema will be
+        disallowed in future attempts.
+        """
+        explain_strategy_none = 'NONE'
+        explain_strategy_procedure = 'PROCEDURE'
+        explain_strategy_statement = 'STATEMENT'
+
+        fns_by_strategy = {
+            explain_strategy_procedure: self._run_explain_procedure,
+            explain_strategy_statement: self._run_explain,
+        }
+
+        plan = None
+        strategy_cache_key = 'explain_strategy:%s' % schema
+        strategy_cache_ttl = 60 * 10
+
+        if not self._can_explain(statement):
+            self.log.debug('Skipping statement which cannot be explained: %s', statement)
+            return None
+
+        if self._expiring_cache.get(strategy_cache_key) == explain_strategy_none:
+            self.log.debug('Skipping statement due to cached collection failure: %s', statement)
+            return None
+
+        exceptions = (pymysql.err.InternalError, pymysql.err.ProgrammingError)
+        non_retryable_errors = frozenset({
```

Comment:
Should this be moved somewhere to the top of the file so it's static?

Reply:
It's not used outside of this function so I think it's better to keep it closer to where it's used.